Remove %guiref and add instead an entry in our Help menu to show its contents

ccordoba12 opened the pull request:
Fixes #78 

----

I'll create a corresponding pull request in the IPython repo to remove `gui_reference` from it.

takluyver commented:
This PR is still registering `%guiref` as a magic. I was initially thinking of a more dramatic change, where we don't involve the kernel in showing it at all, and just make an item in the help menu that brings it up. I know some of the text refers specifically to IPython, though - do you think it's feasible to rewrite these bits so the references to IPython are more examples than prescription? Or is it too tightly wedded to IPython to make any sense for other kernels?

ccordoba12 commented:
> This PR is still registering %guiref as a magic

Yes, but only for `ipykernel`. As far as I know, `ipykernel` is the only kernel that shows `%guiref` as part of its banner. i tried to see if I could remove that line but I didn't understand how to do it.

However, I think it's ok to define `%guiref` when using `ipykernel` because

1. `ipykernel` is able to handle magics
2. People already know about the existence of `%guiref`.

> and just make an item in the help menu that brings it up

I also did that as part of this pull request.

> do you think it's feasible to rewrite these bits so the references to IPython are more examples than prescription?

I removed most of the references to IPython in `gui_reference`, and updated some imports too (like `IPython.core.display` to `IPython.display`). Please review it to see if things can be further improved :-)

takluyver commented:
I think I'd prefer to remove the `%guiref` magic entirely:

- It's only applicable to one frontend, so we shouldn't be showing it in the general banner when starting an IPython kernel. The banner should also show up in the terminal frontend (though it doesn't right now: jupyter/jupyter_console#59), and it's supposed to be independent of the frontend.
- It only works if the Qt console is installed in the same Python that the kernel is running in, which is by no means guaranteed.
- It's one more special case we can get rid of for our own kernel.

The banner that mentions `%guiref` is defined in `IPython.core.usage` and used in `ipykernel.zmqshell`. I'm happy to clear up both of those modules if you want to remove the magic registration in this PR.

ccordoba12 commented:
> It only works if the Qt console is installed in the same Python that the kernel is running in

You're totally right. It makes no sense to maintain `%guiref` because it's not entirely dependent on the kernel.

So please remove its references from `ipykernel` and `IPython.core.usage`, then I'll remove it from here.

----

*Note*: The thing that confused me is that several things about `%guiref` are still present in qtconsole, so I assumed you removed it in a rebase or something. This time I'll make sure to remove all supporting code about it :-)

takluyver commented:
Removed in ipython/ipykernel#101 and ipython/ipython#9231

ccordoba12 commented:
Ok, I think this is ready, unless you have some comments about my changes to `gui_reference`.

ccordoba12 commented:
`%guiref` also appears in this [figure](https://github.com/jupyter/qtconsole/blob/master/docs/source/figs/jn.xhtml), so it should be updated accordingly. Maybe @willingc could help us with that?

takluyver commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+
+This console is designed to emulate the look, feel and workflow of a terminal
+environment, while adding a number of enhancements that are simply not possible
+in a real terminal, such as inline syntax highlighting, true multiline editing,
```
We might want to change the phrasing, because the first two of these (syntax highlighting and multiline editing) are coming to IPython for version 5 (see ipython/ipython#9118). And even inline graphics are [possible](https://github.com/saitoha/PySixel), albeit not supported by most terminal emulators.

willingc replied:
The console is designed to emulate the look, feel and workflow of a terminal environment. Beyond this basic design, the console may also implement functionality not currently found in most terminal emulators. Some examples of these console enhancements are inline syntax highlighting, multiline editing, inline graphics, and others.

takluyver commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+    Hello Jupyter!
+
+    In [2]: for i in range(10):
+       ...:     print i,
```
Let's make our examples Python 3 compliant, to set a good example. And so they don't look like they haven't been updated in years... ;-)

takluyver commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+an ``Out[n]`` prompt only for the last expression in it, while the rest of the
+cell is executed as if it was a script.  An example should clarify this::
+
+    In [3]: x=1  # Hit C-Enter here
```
We should spell out 'Ctrl-', rather than relying on the emacs 'C-' convention.

willingc replied:
+1 to @takluyver's suggestion

takluyver commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+
+- ``C``: Control
+- ``S``: Shift
+- ``M``: Meta (typically the Alt key)
```
Again, my preference is for explicit Ctrl/Shift/Alt in the list below, rather than defining abbreviations.

willingc replied:
+1

takluyver commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+  must quit the pager to get back to the terminal (similar to how a pager such
+  as ``less`` or ``more`` works).
+
+- ``vsplit``: the console is made double-tall, and the pager appears on the
```
Maybe it's just me, but 'double-tall' and 'double-wide' don't sound quite right. I'd say 'double height' and 'double width'.

willingc replied:
Just to clarify, is vsplit defined correctly as is? @takluyver, @ccordoba12 or should it be replaced by hsplit? They seem backwards to me.

Suggestion:
The console is split vertically. The pager appears in the bottom area when needed. You can view the pager's contents while using the terminal.

Similar text for hsplit.

takluyver replied:
Ah, I remember this debate. There is approximately no consensus on which way round the terms 'horizontal split' and 'vertical split' should be applied. To my mind, it does make sense that vsplit puts the pager below, but I can see why other people disagree.

At one point I think we planned to rename them to 'bottom' and 'right', which would avoid this ambiguity. But I guess that never quite happened.

takluyver commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+- ``hsplit``: the console is made double-wide, and the pager appears on the
+  right area when needed.  You can view its contents while using the terminal.
+
+- ``none``: the console never pages output.
```
This isn't very clear: it presumably means that long text that would be paged is displayed like output within the flow of the console, but it could suggest that that output is just discarded.

willingc replied:
Perhaps ``none``: the console displays output without paging.

takluyver commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+Running subprocesses
+====================
+
+The Jupyter QtConsole uses the ``pexpect`` module to run subprocesses
```
It's the IPython kernel that uses Pexpect. But this technical detail is irrelevant to the main point: subprocesses run from the kernel cannot interactively prompt the user for input.

There's a bit of a tricky distinction for new programmers - If you run `foo.py` with something like `%run` or `execfile()`, it is running in the kernel process and *can* get simple input from the user (though full terminal interaction like less or vim is still impossible). But if you run it with `!python`, it's a separate process and cannot prompt the user for any input. I don't know how best to convey this.

ccordoba12 replied:
I also don't know how to address this.

takluyver replied:
I think I'd just remove the first couple of sentences, and start with: "When running a subprocess from the kernel, you can not interact with it as if it was running in a terminal..."

Carreau replied:
OK, I think it's great. If you want to just update this bit and merge (feel free to self merge). Otherwise, I'll merge next time I come by this thread.

takluyver commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+that invokes a pager or expects you to type input into it will block and hang
+(you can kill it with ``Control-C``).
+
+We have provided as magics ``%less`` to page files (aliased to ``%more``),
```
This should make it clear that these magics are an IPython kernel feature, not something the Qt console itself provides.

willingc replied:
The console can use magics provided by the IPython kernel. These magics include:  ``%less`` ...

takluyver commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+=======
+
+The Jupyter QtConsole can now display objects in a variety of formats, including
+PNG and SVG. This is accomplished using the display functions in
```
Again, this should make it clear that it's an example of how to do it from IPython, and that it may be different for other kernels.

Also, I think we can lose the word 'now' in the line above.

willingc replied:
For example, if using the IPython kernel, there are functions available for object display:

takluyver commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+
+The Jupyter QtConsole is capable of displaying matplotlib figures inline, in SVG
+or PNG format.  If started with the ``matplotlib=inline``, then all figures are
+rendered inline automatically (PNG by default).  If started with ``--matplotlib``
```
This flag doesn't even work any more. I think we should just remove this section - the integration with mpl is really an IPython feature, not a Qt console feature.

willingc replied:
I agree lets remove this entire section of "Inline matplotlib graphics"

takluyver commented:
I know most of the things I'm pointing out in the text are not things you changed, but that text is long overdue for an update, and this seems like a good opportunity to go through it. Bad luck!

@willingc , could you also do a pass on this to help make it clearer? Thanks.

willingc commented on qtconsole/jupyter_widget.py:
```diff
@@ -277,13 +276,10 @@ def _handle_kernel_info_reply(self, rep):
 
     def _started_channels(self):
-        """Reimplemented to make a history request and load %guiref."""
+        """Reimplemented to make a history request"""
```
Perhaps a semantic quibble on my part but "reimplemented" feels not quite accurate to me. BaseFrontendWidget is abstract; FrontEndWidget implements the method from the abstract BaseFrontEndWidget; JupyterWidget subclasses from FrontEndWidget.

The docstrings for each method in this section are not consistent. "Override method" or "Redefine method" feel more accurate to me. Now that I'm thinking about it, it may just be best to add a comment here:
```
 #---------------------------------------------------------------------------
 # 'BaseFrontendMixin' abstract interface
 #
 # For JupyterWidget,  override FrontendWidget methods which implement the
 # BaseFrontend Mixin abstract interface
 #---------------------------------------------------------------------------
 ```

willingc commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+
+
+def page_guiref(arg_s=None):
+    """Show a basic reference about QtConsole."""
```
"""Show reference text about QtConsole usage."""

willingc commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+environment.
+
+For single expressions or indented blocks, the console behaves almost like the
+terminal IPython: single expressions are immediately evaluated, and indented
```
IPython terminal

willingc commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+
+The graphical console is capable of true multiline editing, but it also tries
+to behave intuitively like a terminal when possible.  If you are used to
+IPython's old terminal behavior, you should find the transition painless, and
```
... painless. If you learn to use a few basic keybindings, the console provides even greater efficiency.

willingc commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+terminal IPython: single expressions are immediately evaluated, and indented
+blocks are evaluated once a single blank line is entered::
+
+    In [1]: print "Hello Jupyter!"  # Enter was pressed at the end of the line
```
print("Hello Jupyter!")

willingc commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+    In [1]: print "Hello Jupyter!"  # Enter was pressed at the end of the line
+    Hello Jupyter!
+
+    In [2]: for i in range(10):
```
for num in range(10):
    print(num)

willingc commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+
+If you want to enter more than one expression in a single input block
+(something not possible in the terminal), you can use ``Control-Enter`` at the
+end of your first line instead of ``Enter``.  At that point the console goes
```
Huh, didn't know that :-)

willingc commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+(something not possible in the terminal), you can use ``Control-Enter`` at the
+end of your first line instead of ``Enter``.  At that point the console goes
+into 'cell mode' and even if your inputs are not indented, it will continue
+accepting arbitrarily many lines until either you enter an extra blank line or
```
accepting ~~arbitrarily many~~ lines

willingc commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+force execution key, ``Shift-Enter`` (though you can also navigate to the end
+and append a blank line by using ``Enter`` twice).
+
+If you've edited a multiline cell and accidentally navigate out of it with the
```
Is this what is happening:

If you are editing a multiline cell and accidentally navigate out of it using the up or down arrow keys, the console clears the cell and replaces it with the contents of the cell which the up or down arrow key stopped on. If you wish to undo this action, perhaps because of an accidental keypress, use the Undo keybinding, ``Control-z``, to restore the original cell.

willingc commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+=========
+
+The Jupyter QtConsole will show long blocks of text from many sources using a
+builtin pager. You can control where this pager appears with the ``--paging``
```
built-in

willingc commented on qtconsole/usage.py:
```diff
@@ -0,0 +1,221 @@
+command-line flag:
+
+- ``inside`` [default]: the pager is overlaid on top of the main terminal. You
+  must quit the pager to get back to the terminal (similar to how a pager such
```
...similar to how ``less`` or ``more`` pagers behave.)

willingc commented:
@ccordoba12 @takluyver I've gone through and added my comments. Depending on the time window for 4.2 release, it might be good to merge after these changes are made. After that I would be happy to iterate any further changes as needed.

ccordoba12 commented:
Thanks for your review. I addressed all comments, except for one from @takluyver about the use of `subprocess`.

willingc commented:
@ccordoba12 Looks great! I noticed @takluyver had a wording suggestion. Other than that, this is ready to merge.

ccordoba12 commented:
@minrk or @takluyver, I think everything is ready now for you to release 4.2 :-)

willingc commented:
@ccordoba12 :cake: :cookie:

minrk commented:
@ccordoba12 released 4.2! Thanks for your help.

ccordoba12 commented:
Thanks a lot Min!!

El 23/02/16 a las 03:04, Min RK escribió:
>
> @ccordoba12 <https://github.com/ccordoba12> released 4.2! Thanks for 
> your help.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/jupyter/qtconsole/pull/98#issuecomment-187594623>.
>

temok-mx commented:
Dear All,

Wishes for the QtConsole:
1)   Enable searches like in Emacs (`control-s`) or like in the `less` command (slash '/').
This would be useful in situations like when browsing the QtConsole Help.

2)   Enable the display of the QtConsole Help from a command in the QtConsole (in addition or in replacement of using the mouse to click on the Help menu).

Cheers,
Temok

ccordoba12 commented:
Regarding your points @temok-mx:

1) This is implemented in Spyder, i.e. if you press Ctrl+F you get a search dialog to easily look for something in Qtconsole content.

2) Previously Qtconsole had a command called `%guiref` to do precisely this, but it was removed by this PR. The reasoning behind this decision is described in https://github.com/jupyter/qtconsole/pull/98#issuecomment-184202984

3) That's a good idea, but someone will have to implement it. You could open a new issue about it :-)

temok-mx commented:
Thanks Carlos,

I just posted an issue for point 3):  #122 
Saludos,
Temok
